File: week07/19/P4.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

bool partition(int K, int mid, int A[], int N) {
    int temp = 0;
    int count = 1;
    for (int i=0; i<N; i++) {
        if (temp+A[i] <= mid) temp += A[i];
        else {
            count++;
            temp = A[i];
        }
    }
    if (count > K) return false;
    return true;
}

int solution(int K, int M, int A[], int N) {
    int upper = 0;
    for (int i=0; i<N; i++) upper += A[i];
    int lower = M;
    int result = upper;
    while (lower <= upper) {
        int mid = (upper + lower) / 2;
        if (partition(K, mid, A, N)) {
            result = mid;
            upper = mid - 1;
        } else lower = mid + 1;
    }
    return result;
}

int main() {
    int A[] = {2, 1, 5, 1, 2, 2, 2};
    int AN = sizeof(A)/sizeof(int);
    printf("%d\n", solution(3, 5, A, AN));

    int B[] = {5, 3};
    int BN = sizeof(B)/sizeof(int);
    printf("%d\n", solution(1, 10, B, BN));

    int C[] = {3, 4, 5, 6, 7, 8};
    int CN = sizeof(C)/sizeof(int);
    printf("%d\n", solution(4, 10, C, CN));

    return 0;
}

File: week07/19/P4.py
# mid 값을 넘지 않고 K개 이내로 분할이 가능한 경우 최대값 출력, 아니면 -1 출력
def partition(K, mid, A):
    temp = 0
    part = []
    for i in A:
        if (temp+i <= mid):
            temp += i
        else:
            part.append(temp)
            temp = i
    part.append(temp)
    if (len(part) <= K):
        return max(part)
    return -1

def solution(K, M, A):
    upper = sum(A)
    lower = max(A)
    answer = []
    while (lower <= upper):
        mid = (upper + lower) // 2
        answer.append(partition(K, mid, A))
        if (answer[-1] != -1): # mid 값을 넘지 않고 K개 이내로 분할이 가능한 경우
            upper = mid - 1
        else:
            lower = mid + 1
            answer.pop()
    if (answer == []):
        return lower
    return min(answer)

if __name__ == "__main__":
    A = [2, 1, 5, 1, 2, 2, 2]
    print(solution(3, 5, A))
    B = [5, 3]
    print(solution(1, 10, B))
    C = [3, 4, 5, 6, 7, 8]
    print(solution(4, 10, C))